aten/src/ATen/native/cuda/ForeachUnaryOp.cu

```diff
@@ -180,150 +204,164 @@ void foreach_unary_op_bfloat16_(TensorList tensors) {
+    __device__ T operator()(T t) const { return std::NAME(t); }                    \
+};                                                                                 \
+                                                                                   \
+std::vector<Tensor> foreach_tensor_##NAME##_cuda(TensorList tensors) {             \
+    check_foreach_api_restrictions(tensors);                                       \
+    if (!can_use_fast_route(tensors)) {                                            \
+        return at::native::foreach_tensor_##NAME##_slow(tensors);                  \
+    }                                                                              \
+                                                                                   \
+                                                                                   \
+    return floating_half<NAME1>(tensors);                                          \
+}                                                                                  \
+                                                                                   \
+void foreach_tensor_##NAME##_cuda_(TensorList tensors) {                           \
+    check_foreach_api_restrictions(tensors);                                       \
+    if (!can_use_fast_route(tensors)) {                                            \
+        at::native::foreach_tensor_##NAME##_slow_(tensors);                        \
+    }                                                                              \
+                                                                                   \
+    floating_half_<NAME1>(tensors);                                                \
 }
 
-FOREACH_UNARY_OP(ceil, Ceil);
-FOREACH_UNARY_OP(erfc, Erfc);
-FOREACH_UNARY_OP(expm1, Expm1);
-FOREACH_UNARY_OP(floor, Floor);
-FOREACH_UNARY_OP(lgamma, Lgamma);
-
-FOREACH_UNARY_OP_BFLOAT16(log1p, Log1p);
-FOREACH_UNARY_OP_BFLOAT16(erf, Erf);
-
-FOREACH_UNARY_OP_COMPLEX(acos, Acos);
-FOREACH_UNARY_OP_COMPLEX(asin, Asin);
-FOREACH_UNARY_OP_COMPLEX(atan, Atan);
-FOREACH_UNARY_OP_COMPLEX(cosh, Cosh);
-FOREACH_UNARY_OP_COMPLEX(tan, Tan);
-FOREACH_UNARY_OP_COMPLEX(sin, Sin);
-FOREACH_UNARY_OP_COMPLEX(sinh, Sinh);
-
-FOREACH_UNARY_OP_COMPLEX_BFLOAT16(exp, Exp);
-FOREACH_UNARY_OP_COMPLEX_BFLOAT16(sqrt, Sqrt);
-FOREACH_UNARY_OP_COMPLEX_BFLOAT16(cos, Cos);
-FOREACH_UNARY_OP_COMPLEX_BFLOAT16(tanh, Tanh);
-FOREACH_UNARY_OP_COMPLEX_BFLOAT16(log, Log);
-FOREACH_UNARY_OP_COMPLEX_BFLOAT16(log10, Log10);
-FOREACH_UNARY_OP_COMPLEX_BFLOAT16(log2, Log2);
+FLOATING_HALF(erfc, Erfc);
+FLOATING_HALF(expm1, Expm1);
+FLOATING_HALF(lgamma, Lgamma);
+FLOATING_HALF(trunc, Truncf);
+FLOATING_HALF(floor, Floor);
+FLOATING_HALF(ceil, Ceil);
+
+template<typename T>
+struct Log1p {
+    __device__ T operator()(T t) const { return std::log1p(t); }
+};
+
+template<typename T>
+struct Erf {
+    __device__ T operator()(T t) const { return std::erf(t); }
+};
```

Comment:
We didn't have to write these out before, but now we do :(

Reply:
The suggestion here https://github.com/pytorch/pytorch/pull/49248#discussion_r545216645 offers a way to fix this; in particular, it creates two macros: one with a way to register a custom (non-std-based) Functor, and one that just autogenerates a std:: based functor.